playback/src/audio_backend/alsa.rs

```diff
@@ -228,54 +236,96 @@ impl Sink for AlsaSink {
-            // If recovery fails then forward that error back to player.
-            warn!(
-                "Error writing from AlsaSink buffer to PCM, trying to recover {}",
-                err
-            );
-            pcm.try_recover(err, false).map_err(|e| {
-                io::Error::new(
-                    io::ErrorKind::Other,
-                    format!(
-                        "Error writing from AlsaSink buffer to PCM, recovery failed {}",
-                        e
-                    ),
-                )
-            })?
+    fn configure_period_buffer(&mut self, bytes_per_period: usize) {
+        // Vec growth strategies can and have changed.
+        // Capacity may not always be trustworthy in the future.
+        // Use bytes_per_period to make sure that no matter what
+        // we give the PCM a period's worth of audio.
+        // Don't depend on len == capacity.
+        self.bytes_per_period = bytes_per_period;
+        // As noted above Vec growth strategies are not guaranteed
+        // so we over allocate initially so that our Vec will never
+        // actually reach full capacity thus (hopefully) avoiding
+        // future reallocations. Nearest (truncated) KiB + 2 KiB.
+        let over_allocation = ((bytes_per_period / KIB) * KIB) + TWO_KIB;
+        // If the capacity is greater than we want try to shrink it
+        // to it's current len (which should be zero) before trying
+        // to set the capacity with reserve_exact.
+        if self.period_buffer.capacity() > over_allocation {
+            self.period_buffer.shrink_to_fit();
+        }
+        // This does nothing if the capacity is already sufficient.
+        // Len should always be zero, but for the sake of being thorough...
+        self.period_buffer
+            .reserve_exact(over_allocation - self.period_buffer.len());
+    }
+
+    fn process_bytes(&mut self, data: &[u8], start_index: usize) -> io::Result<()> {
+        let space_left = self.bytes_per_period - self.period_buffer.len();
+        let data_size = data.len();
+        let end_index = min(space_left, data_size);
+        self.period_buffer
+            .extend_from_slice(&data[start_index..end_index]);
+        if self.period_buffer.len() == self.bytes_per_period {
+            self.write_buf()?;
         }
+        if end_index < data_size {
+            self.process_bytes(data, end_index)?;
+        }
+        Ok(())
+    }
+
+    fn drain_period_buffer(&mut self) -> io::Result<()> {
+        // Always give the PCM a period's worth of audio
+        // even when draining the buffer. This makes for
+        // more consistent stop to start latency
+        // (pause to play, manually changing tracks)
+        // and helps to prevent jarring, "chopped off" sounding
+        // transitions when changing tracks manually.
+        // Overall resulting in a slightly smoother user experience.
+        self.period_buffer.resize(self.bytes_per_period, u8::MIN);
```

Comment:
That's a neat trick, and now that I Googled it, indeed a issue with how Alsa works under the hood. Two points:
1. Just `0` instead of `u8::MIN` seems more obvious
2. `resize` could again push it beyond period boundaries. Instead you could consider using `append` with a newly constructed `vec![0; capacity - len]`.

Reply:
> Just 0 instead of u8::MIN seems more obvious

That's fine. I can do that.

> resize could again push it beyond period boundaries. Instead you could consider using append with a newly constructed vec![0; capacity - len].

The docs say:

> Resizes the Vec in-place so that len is equal to new_len.

> If new_len is greater than len, the Vec is extended by the difference, with each additional slot filled with value. If new_len is less than len, the Vec is simply truncated.

What you're looking for now that we've decided that capacity can be trusted is:

```rust
self.period_buffer.resize(self.period_buffer.capacity(), 0);
```

That would zero fill the remainder of the buffer. It would not make it grow any larger.

Reply:
👍

Reply:
Done.